### PR TITLE
Separate GridSample Coordinate Types from Data Types

### PR DESCRIPTION
**Pull request by leimao**
Follow up PR #96586

The operator math can be used for two situations in GridSample forward call:
1. Fractional coordinate computation.
2. Output activation interpolation (because it involves accumulation).

The operator math can also be used for two situations in GridSample backward call:
1. Fractional coordinate computation.
2. Input activation gradient computation.

Currently, on PyTorch main, the GridSample forward call uses operator math for both 1 and 2. However, the GridSample backward call does not use operator math at all for 1 and 2. 

Ideally, the operator math usages should be consistent between the forward call and the backward call, especially for the fractional coordinate computation. Otherwise, the input gradient will be updated for "incorrect" pixels from the input tensor.

The input activation gradient computation of the GridSample backward call can hardly use operator math because it uses atomic operations for gradient accumulation and this can hardly be replaced without significantly changing the implementation.

In this patch, I updated the GridSample backward call so that it uses operator math for the fractional coordinate computation and it will be consistent with the forward call.

### CODE REVIEW

**Comment by leimao** [comment]
@albanD Could you please assign someone to carefully review this PR? Thank you.

**Comment by leimao** [comment]
@mikaylagawarecki Could you please take a look when you get a chance? Thanks.

File: aten/src/ATen/native/GridSampler.h [hunk 1]
```diff
@@ -38,16 +38,16 @@ static inline scalar_t grid_sampler_unnormalize(scalar_t coord, int64_t size,
 // except that it also returns the `d output / d input` via pointer argument
 // `grad_in`.
 // This is useful in the backward pass of grid_sampler.
-template <typename scalar_t>
-static inline scalar_t grid_sampler_unnormalize_set_grad(scalar_t coord, int64_t size,
-                                                         bool align_corners, scalar_t *grad_in) {
+template <typename coord_scalar_t, typename grad_scalar_t>
```
Comment (leimao) [inline aten/src/ATen/native/GridSampler.h:41]: Might consider renaming `grad_scalar_t` to `data_scalar_t` to make the changes consistent.

File: aten/src/ATen/native/cuda/GridSampler.cu [hunk 2]
```diff
@@ -393,10 +396,10 @@ namespace {
...
-            safe_add_2d(grad_input.data, iy_nw, ix_nw, gInp_sH, gInp_sW, inp_H, inp_W, nw * gOut, NC_offset, grad_input_memory_span);
-            safe_add_2d(grad_input.data, iy_ne, ix_ne, gInp_sH, gInp_sW, inp_H, inp_W, ne * gOut, NC_offset, grad_input_memory_span);
-            safe_add_2d(grad_input.data, iy_sw, ix_sw, gInp_sH, gInp_sW, inp_H, inp_W, sw * gOut, NC_offset, grad_input_memory_span);
-            safe_add_2d(grad_input.data, iy_se, ix_se, gInp_sH, gInp_sW, inp_H, inp_W, se * gOut, NC_offset, grad_input_memory_span);
+            safe_add_2d(grad_input.data, iy_nw, ix_nw, gInp_sH, gInp_sW, inp_H, inp_W, static_cast<scalar_t>(nw * gOut), NC_offset, grad_input_memory_span);
+            safe_add_2d(grad_input.data, iy_ne, ix_ne, gInp_sH, gInp_sW, inp_H, inp_W, static_cast<scalar_t>(ne * gOut), NC_offset, grad_input_memory_span);
+            safe_add_2d(grad_input.data, iy_sw, ix_sw, gInp_sH, gInp_sW, inp_H, inp_W, static_cast<scalar_t>(sw * gOut), NC_offset, grad_input_memory_span);
+            safe_add_2d(grad_input.data, iy_se, ix_se, gInp_sH, gInp_sW, inp_H, inp_W, static_cast<scalar_t>(se * gOut), NC_offset, grad_input_memory_span);
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/cuda/GridSampler.cu:402]: noob question: why do we `use static_cast<scalar_t>`  here but not on [L495](https://github.com/pytorch/pytorch/pull/120543/files#diff-bb063818c22c14573e6d553e9ba46402cb866917d5d93c7ac0a0831fa4e62c49R495)?

Do we need a `static_cast<scalar_t>(gOut * x_coeffs[i] * y_coeffs[j])` on line 495 as well, since that has `gOut (scalar_t), x_coeffs[i] (opmath_t),  y_coeffs[j] (opmath_t)`?

**Review by mikaylagawarecki — COMMENTED** [review]
Thanks!

Just to make sure I interpreted the PR accurately, this PR

1) Makes some fixes to the forward kernels where accumulation output was not getting `static_cast`-ed from `opmath_t` to `scalar_t` 
2) Adds templating to decouple the type of the gradients from the type of the coordinates since we want coordinate calculation to use `opmath_t` but not gradient computation (since it uses atomics). This allows the coordinates calculated in backward to be consistent with forward.

Does this sound right to you?

Separately, do you think we could add a test to verify the updated behavior? (Perhaps a case where the gradient would have been incorrect before due to discrepancies in coordinate calculation between fw and bw but now passes after this fix?)

**Comment by leimao** [comment]
> Thanks!
> 
> Just to make sure I interpreted the PR accurately, this PR
> 
> 1. Makes some fixes to the forward kernels where accumulation output was not getting `static_cast`-ed from `opmath_t` to `scalar_t`
> 2. Adds templating to decouple the type of the gradients from the type of the coordinates since we want coordinate calculation to use `opmath_t` but not gradient computation (since it uses atomics). This allows the coordinates calculated in backward to be consistent with forward.
> 
> Does this sound right to you?
> 
> Separately, do you think we could add a test to verify the updated behavior? (Perhaps a case where the gradient would have been incorrect before due to discrepancies in coordinate calculation between fw and bw but now passes after this fix?)

Thank you @mikaylagawarecki for your timely review. 

For the point 1, actually without my explicit `static_cast<scalar_t>`, the output will be implicitly casted to `scalar_t` because the output tensor data type was declared to be `scalar_t`. Having an explicit `static_cast<scalar_t>` just makes the code more readable.

For the point 2, your understanding is consistent with the motivation of my PR. Otherwise, the pixels sampled for interpolation in the forward propagation might be different from the pixels being updated in the back propagation.

I do think we need a unit test for guarding this. Otherwise, in the future, if someone changed the code and alter the behavior, we would not be alerted. However, I have not thought about how to create such a unit test yet. I am creating this PR to first collect some feedback from the community. If the community thinks there is no need for this PR, of course I would not be spending time on creating the unit test. If the community thinks this PR is necessary, I can try adding a unit test to this PR later.

File: aten/src/ATen/native/cuda/GridSampler.cu (hunk 2 above)
Comment (leimao) [reply aten/src/ATen/native/cuda/GridSampler.cu:402]: Notice the `safe_add_2d` function signature,
```
template<typename scalar_t, typename index_t>
static __forceinline__ __device__
void safe_add_2d(scalar_t *data, int h, int w,
                 int sH, int sW, int H, int W,
                 scalar_t delta,
                 const index_t NC_offset,
                 const index_t memory_span)
```
In `safe_add_2d`, we relied on the compiler to infer what the template variable types are from the variable it received. That's why I explicitly casted the `nw * gOut` from `opmath_t` to `scalar_t`. Otherwise the compiler would be confused.

On line 495, the function is `add_value_bounded` and the function signature is
```
template<typename data_scalar_t, typename coord_scalar_t, typename index_t>
static __forceinline__ __device__
void add_value_bounded(
    data_scalar_t* data, coord_scalar_t x, coord_scalar_t y, int W, int H, int sW, int sH,
    data_scalar_t delta,
    GridSamplerPadding padding_mode,
    bool align_corners,
    const index_t NC_offset,
    const index_t memory_span)
```
We already specified the `data_scalar_t` to be `scalar_t` in the function call `add_value_bounded<scalar_t>`, therefore, the compiler would not be confused and it will implicitly cast the `gOut * x_coeffs[i] * y_coeffs[j]` from `opmath_t` to `scalar_t`. But of course, for code readability, we want everything to be explicit and we can have a `static_cast<scalar_t>` there.

Comment (mikaylagawarecki) [reply aten/src/ATen/native/cuda/GridSampler.cu:402]: That makes a lot of sense, thank you for the detailed explanation!

**Comment by mikaylagawarecki** [comment]
Thanks for your response!

>  I am creating this PR to first collect some feedback from the community. If the community thinks there is no need for this PR, of course I would not be spending time on creating the unit test. If the community thinks this PR is necessary, I can try adding a unit test to this PR later.

Curious how you intend to get feedback from the community on this? Are you looking for feedback from a maintenance perspective or from the wider community?

One metric we tend to refer to might be reactions/activities on issues (so it might be useful to create an accompanying issue for this).

However, in my opinion, the error that you are fixing seems like a silent correctness error, which from a maintenance perspective, seems more like a "should fix" to me :) Will bring this up in our triage discussion if that sounds good to you!

**Comment by leimao** [comment]
> Thanks for your response!
> 
> > I am creating this PR to first collect some feedback from the community. If the community thinks there is no need for this PR, of course I would not be spending time on creating the unit test. If the community thinks this PR is necessary, I can try adding a unit test to this PR later.
> 
> Curious how you intend to get feedback from the community on this, and what kind of feedback you are looking for. Are you looking for feedback from a maintenance perspective or from the wider community?
> 
> One metric we tend to refer to might be reactions/activity on issues reporting the bug (so it might be useful to create an accompanying issue for this).
> 
> However, in my opinion, the problem that you are fixing seems like a silent correctness error, which from a maintenance perspective, is a "must fix" :) Will bring this up in our triage discussion if that sounds good to you!

@mikaylagawarecki you are also part of the community :) and your opinion is part of the community feedback :)
There is no rush to merge this PR. But please keep me updated if there are some conclusions from your internal discussions. Thanks.

**Comment by mikaylagawarecki** [comment]
@leimao Hey, apologies I forgot to convey this but fyi our conclusion on this was that it is indeed a silent correctness error, which should be fixed and adding testing here would be great

**Comment by leimao** [comment]
> @leimao Hey, apologies I forgot to convey this but fyi our conclusion on this was that it is indeed a silent correctness error, which should be fixed and adding testing here would be great

Thanks @mikaylagawarecki. I will try finding some time in the next few weeks to come up with some unit tests. Might bother you then.
